src/wallet/wallet2.cpp

```diff
@@ -5373,34 +5373,91 @@ uint64_t wallet2::import_key_images(const std::vector<std::pair<crypto::key_imag
-  }
-  LOG_PRINT_L1("Total: " << print_money(spent) << " spent, " << print_money(unspent) << " unspent");
+  if(check_spent)
+  {
+    m_daemon_rpc_mutex.lock();
+    bool r = epee::net_utils::invoke_http_json("/is_key_image_spent", req, daemon_resp, m_http_client, rpc_timeout);
+    m_daemon_rpc_mutex.unlock();
+    THROW_WALLET_EXCEPTION_IF(!r, error::no_connection_to_daemon, "is_key_image_spent");
+    THROW_WALLET_EXCEPTION_IF(daemon_resp.status == CORE_RPC_STATUS_BUSY, error::daemon_busy, "is_key_image_spent");
+    THROW_WALLET_EXCEPTION_IF(daemon_resp.status != CORE_RPC_STATUS_OK, error::is_key_image_spent_error, daemon_resp.status);
+    THROW_WALLET_EXCEPTION_IF(daemon_resp.spent_status.size() != signed_key_images.size(), error::wallet_internal_error,
+      "daemon returned wrong response for is_key_image_spent, wrong amounts count = " +
+      std::to_string(daemon_resp.spent_status.size()) + ", expected " +  std::to_string(signed_key_images.size()));
 
+    spent = 0;
+    unspent = 0;
+    for (size_t n = 0; n < daemon_resp.spent_status.size(); ++n)
+    {
+      transfer_details &td = m_transfers[n];
+      uint64_t amount = td.amount();
+      td.m_spent = daemon_resp.spent_status[n] != COMMAND_RPC_IS_KEY_IMAGE_SPENT::UNSPENT;
+      if (td.m_spent)
+        spent += amount;
+      else
+        unspent += amount;
+      LOG_PRINT_L2("Transfer " << n << ": " << print_money(amount) << " (" << td.m_global_output_index << "): "
+          << (td.m_spent ? "spent" : "unspent") << " (key image " << req.key_images[n] << ")");
+    }
+  }
+  else
+  {
+    spent = 0;
+    unspent = 0;
+    for(size_t i = 0; i < m_transfers.size(); ++i)
+    {
+      transfer_details &td = m_transfers[i];
+      uint64_t amount = td.amount();
+      if (td.m_spent)
+        spent += amount;
+      else
+        unspent += amount;
+    }
+  }
+  MDEBUG("Total: " << print_money(spent) << " spent, " << print_money(unspent) << " unspent");
   return m_transfers[signed_key_images.size() - 1].m_block_height;
 }
+wallet2::payment_container wallet2::export_payments() const
+{
+  payment_container payments;
+  for (auto const p : m_payments)
+  {
+    payments.emplace(p);
+  }
+  return payments;
+}
+void wallet2::import_payments(const payment_container &payments)
+{
+  for (auto const &p : payments)
+  {
+    m_payments.emplace(p);
+  }
+}
+void wallet2::import_payments_out(const std::list<std::pair<crypto::hash,wallet2::confirmed_transfer_details>> &confirmed_payments)
+{
+  for (auto const &p : confirmed_payments)
+  {
+    m_confirmed_txs.emplace(p);
+  }
+}
+
+std::vector<crypto::hash> wallet2::export_blockchain() const
+{
+  std::vector<crypto::hash> bc;
+  for (auto const &b : m_blockchain)
+  {
+    bc.push_back(b);
+  }
+  return bc;
+}
+
+void wallet2::import_blockchain(const std::vector<crypto::hash> &bc)
+{
+  for (auto const &b : bc)
+  {
+    m_blockchain.push_back(b);
+  }
+}
```

Comment:
This needs to clear m_blockchain first, and check the first hash matches the genesis block, and update m_local_blockchain_height at the end.